### PR TITLE
Rework the relation interface to be more flexible and robust

### PR DESCRIPTION
**Pull request by dstathis**
After discussions about how database relations should request new databases, I have reworked the charm library.

### CODE REVIEW

File: lib/charms/cassandra_k8s/v0/cassandra.py [hunk 1]
```diff
@@ -46,11 +45,39 @@ def new_func(self, *args, **kwargs):
     return new_func
 
 
-class CQLConsumer(ConsumerBase):
+class CassandraConsumerError(Exception):
```
Comment (balbirthomas) [inline lib/charms/cassandra_k8s/v0/cassandra.py:48]: Isn't this exception name so generic that it does not indicate the type of error that caused it ? The exception is thrown when database names are not unique or the database name is too long. How is the code trapping this error to decide which of these two errors was raised and thereby decide how to deal with the error ? Why not create tow exceptions instead something like `DuplicateDatabaseNames`  and `OversizedDatabaseName` ?

File: lib/charms/cassandra_k8s/v0/cassandra.py [hunk 2]
```diff
@@ -75,27 +102,26 @@ def databases(self, rel_id=None):
...
-    def request_databases(self, num_databases, rel_id=None):
-        """Request n databases"""
-        if not self.charm.unit.is_leader():
-            return
-
-        rel = self.framework.model.get_relation(self.relation_name, rel_id)
-
-        rel.data[self.charm.app]['requested_databases'] = str(num_databases)
+        if name_suffix:
+            name_suffix = "_{}".format(name_suffix)
```
Comment (balbirthomas) [inline lib/charms/cassandra_k8s/v0/cassandra.py:124]: If you are allow the caller to specify part of the database name, why not the whole name ? Prepending an underscore does not make the database name unique. Prepending something like `juju_<model-name>_<app-name>_<relation-id>` might.

File: lib/charms/cassandra_k8s/v0/cassandra.py [hunk 3]
```diff
@@ -75,27 +102,26 @@ def databases(self, rel_id=None):
...
+        db_name = "juju_db_{}_{}{}".format(sanitize_name(self.charm.model.name), sanitize_name(self.charm.app.name), sanitize_name(name_suffix))
+        if len(db_name) > 48:
```
Comment (balbirthomas) [inline lib/charms/cassandra_k8s/v0/cassandra.py:127]: A comment here explaining the reason for this restriction to 48 characters might be useful.

**Review by balbirthomas — APPROVED** [review]

File: lib/charms/cassandra_k8s/v0/cassandra.py (hunk 2 above)
Comment (dstathis) [reply lib/charms/cassandra_k8s/v0/cassandra.py:124]: We want to have juju in front for anything juju generated. The idea of model and app name is to guarantee that the name is unique from other applications. The idea was basically that if you only want one database, you don't have to do anything for uniqueness.

Comment (dstathis) [reply lib/charms/cassandra_k8s/v0/cassandra.py:124]: oh I think you also may not have seen that I prepend `juju_<model-name>_<app-name>` a line or two later.

File: lib/charms/cassandra_k8s/v0/cassandra.py (hunk 1 above)
Comment (dstathis) [reply lib/charms/cassandra_k8s/v0/cassandra.py:48]: Done

File: lib/charms/cassandra_k8s/v0/cassandra.py (hunk 3 above)
Comment (dstathis) [reply lib/charms/cassandra_k8s/v0/cassandra.py:127]: Done

**Pull request closed without merging**